habitat-lab/habitat/datasets/rearrange/modify_episodes_for_object_rearrange.py

```diff
@@ -0,0 +1,427 @@
+        # for YCB objects
+        return obj_name[4:]
+    elif obj_name in obj_category_mapping:
+        return obj_category_mapping[obj_name]
+    else:
+        raise ValueError(f"Cannot find matching category for {obj_name}")
+
+
+def get_obj_rec_cat_in_eps(
+    episode, obj_category_mapping=None, rec_category_mapping=None
+):
+    """
+    Returns the category names for the object to be picked up, start and goal receptacles
+    """
+    obj_cat = get_obj_category(
+        list(episode["info"]["object_labels"].keys())[0].split(":")[0][:-1],
+        obj_category_mapping=obj_category_mapping,
+    )
+    start_rec_cat = get_rec_category(
+        episode["target_receptacles"][0][0].split(":")[0],
+        rec_category_mapping=rec_category_mapping,
+    )
+    goal_rec_cat = get_rec_category(
+        episode["goal_receptacles"][0][0].split(":")[0],
+        rec_category_mapping=rec_category_mapping,
+    )
+    return obj_cat, start_rec_cat, goal_rec_cat
+
+
+def read_obj_category_mapping(filename):
+    df = pd.read_csv(filename)
+    df["category"] = df["clean_category"].apply(lambda x: x.replace(" ", "_"))
+    return dict(zip(df["name"], df["category"]))
+
+
+def get_cats_list(
+    train_episodes_file, obj_category_mapping=None, rec_category_mapping=None
+):
+    """
+    Extracts the lists of object and receptacle categories from episodes and returns name to id mappings
+    """
+    episodes = json.load(gzip.open(train_episodes_file))
+    obj_cats, rec_cats = set(), set()
+    o = list()
+    for episode in episodes["episodes"]:
+        obj_cat, start_rec_cat, goal_rec_cat = get_obj_rec_cat_in_eps(
+            episode,
+            obj_category_mapping=obj_category_mapping,
+            rec_category_mapping=rec_category_mapping,
+        )
+        obj_cats.add(obj_cat)
+        rec_cats.add(start_rec_cat)
+        rec_cats.add(goal_rec_cat)
+        o.append(obj_cat)
+    obj_cats = sorted(obj_cats)
+    rec_cats = sorted(rec_cats)
+    obj_to_id_mapping = {cat: i for i, cat in enumerate(obj_cats)}
+    rec_to_id_mapping = {cat: i for i, cat in enumerate(rec_cats)}
+    return obj_to_id_mapping, rec_to_id_mapping
+
+
+def collect_receptacle_goals(sim, rec_category_mapping=None):
+    recep_goals = []
+    receptacles = find_receptacles(sim)
+    rom = sim.get_rigid_object_manager()
+    aom = sim.get_articulated_object_manager()
+
+    for receptacle in receptacles:
+        if receptacle.parent_object_handle is None:
+            object_id = -1
+        elif aom.get_library_has_handle(receptacle.parent_object_handle):
+            object_id = aom.get_object_id_by_handle(
+                receptacle.parent_object_handle
+            )
+        elif rom.get_library_has_handle(receptacle.parent_object_handle):
+            object_id = rom.get_object_id_by_handle(
+                receptacle.parent_object_handle
+            )
+        else:
+            object_id = -1
+        pos = receptacle.get_surface_center(sim)
```

Comment:
Why do we need the center position of the receptacle?

Reply:
> I think it makes sense to change this to be with respect to a random position on the receptacle for better robustness.
Aren't we already uniformly sampling a start position within 2m of the receptacle center position. From what I understand, you are proposing sampling of a position on the receptacle and then using that position as a center of a circle from which you will uniformly sample a point? I think this is going to bias the distribution of starting points. Do you think it is beneficial to have this bias?

Reply:
Yes, you are right. For pick to succeed, currently we require the nav to end at a point 2m from the center of receptacle. I was thinking that sampling robot position from a point 2m from any point on receptacle would make the Pick more robust - the nav can stop at a point that is close to any point on the receptacle. I think this would be the case when we switch to using viewpoints for Nav's success. But this may make the Pick training harder. For now, I am adding a config param to support and experiment with both.

I don't think this would bias the distribution of starting points because sampling a point on the receptacle is expected to be random.

Reply:
> I don't think this would bias the distribution of starting points because sampling a point on the receptacle is expected to be random.

Lets say you have a square table of size 2x2 and you uniformly sample a point on the surface of this table, called p1. Once this point has been sampled, you create a circle of radius 2m and sample a point (p2) from the circle, again uniformly. The maximum distance of p2 from the table center is going to be 4m and p2 will only be sampled if the p1 was sampled at the edge of the table. 

On the other hand, if p2 is very close to the edge of the table, it could have been sampled from a lot of different p1s. So p2 will definitely have a bias.

Reply:
Oh, got it now. Yes, p2 will be biased. Will keep the current implementation.